### PR TITLE
differentiate BusContext and BusSession

### PR DESCRIPTION
**Pull request by timbussmann**
on a discussion with @danielmarbach @SzymonPobiega and @hmemcpy we tried to find a way to avoid damage caused by users binding `IBusContext` to `Endpoint.CreateBusContext` to enable DI of the bus context. This is potentially dangerous because it would inject a context-unaware `IBusContext` if requested from within the pipeline.

We can lower potential damage by such scenarios by having `IBusSession` **and** `IBusContext` where
* `IBusContext` is the context-aware bus access usable from within the pipeline (behaviors, handlers, sagas, ...)
* `IBusSession` is the context-unaware bus access for access outside of the pipeline created by `endpoint.CreateBusSession()`.

This results in 
* Clear differentiation of the two usage scenarios by having complete independent interfaces.
* Users can easily bind `IBusSession` for DI usage outside the pipeline.
* If users absolutely require DI of BusContext within the pipeline, they can bind that too using a custom behavior without interfering with BusSession bindings.
* We can scope certain operations to BusSession or BusContext. E.g. Callbacks or Tasks.
* We have nearly identical interfaces & extensions which need to be in sync.

The current PR limits the possibility of scheduling tasks to BusSession which means that tasks can no longer be scheduled from message handlers / behaviors. Thoughts on that? (Also consider non-DTC configurations)

### CODE REVIEW

**Comment by timbussmann** [comment]
@Particular/tf-handlercontext @Particular/nservicebus please review

**Comment by hmemcpy** [comment]
:+1:

**Comment by SzymonPobiega** [comment]
@timbussmann I think the scheduling gotcha is fine. When you are in a handler, you can send a deferred message to yourself.

**Comment by timbussmann** [comment]
I mean it wouldn't be much effort to support scheduling again. But as you mentioned, there are probably better/safer approaches to it.

File: src/NServiceBus.Core/BusContext.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@ namespace NServiceBus
     using NServiceBus.Pipeline;
     using NServiceBus.Unicast;
 
-    class BusContext : IBusContext
+    class BusContext : IBusContext, IBusSession
```
Comment (danielmarbach) [inline src/NServiceBus.Core/BusContext.cs]: I think we can't/should implement both interfaces on `BusContext`. We could run into problem when a container registers by convention bindings for all public interfaces.

File: src/NServiceBus.Core/Features/FeatureStartupTaskController.cs [hunk 2]
```diff
@@ -16,15 +16,15 @@ public FeatureStartupTaskController(string name, Func<IBuilder, FeatureStartupTa
...
             instance = factory(builder);
-            return instance.PerformStartup(busContext);
+            return instance.PerformStartup(busSession);
         }
 
-        public async Task Stop(IBusContext busContext)
+        public async Task Stop(IBusSession IBusSession)
         {
-            await instance.PerformStop(busContext);
+            await instance.PerformStop(IBusSession);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Features/FeatureStartupTaskController.cs]: `ConfigureAwait(false)`

File: src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs [hunk 3]
```diff
@@ -47,8 +47,8 @@ public IBusContext CreateBusContext()
         StartAndStoppablesRunner startAndStoppablesRunner;
         FeatureRunner featureRunner;
         IBuilder builder;
+        IBusContextFactory busContextFactory;
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs]: `IBusSessionFactory`?

File: src/NServiceBus.Core/BusContext.cs (hunk 1 above)
Comment (timbussmann) [inline src/NServiceBus.Core/BusContext.cs]: yep, absolutely right, going to fix :+1:

**Comment by timbussmann** [comment]
rebased

**Comment by danielmarbach** [comment]
:+1: 

Can you make sure you update the doco accordingly, asap?

**Pull request merged**